src/scripts/hunt-filter/interfaces.ts

```diff
@@ -0,0 +1,60 @@
+import { IntakeMessage } from "../types/mhct";
+
+/**
+ * Describes a way to validate a pre and post object
+ */
+export interface IRule<K> {
+    /**
+     * Rule name. Used for logging.
+     */
+    readonly name: string;
+    /**
+     * Check if two objects are valid between each other.
+     * @param pre The pre-hunt object
+     * @param post The post-hunt object
+     * @returns true if valid, otherwise false
+     */
+    isValid(pre: K, post: K): boolean;
+}
+
+/**
+ * Provides a rule for a specific property of T
+ */
+export interface IFilteredRule<T> extends IRule<T> {
+    readonly property: (keyof T);
+}
+
+/**
+ * Contract for implementing exemptions of IntakeMessage
+ */
+export interface IMessageExemption {
+    /**
+     * 
+     */
+    readonly name: string;
+    /**
+     * When this property name is invalid on an intake message, 
+     * this rule will try to provide exemptions.
+     */
+    readonly property: (keyof IntakeMessage);
+    /**
+     * Get exemptions for the given pre and post messages
+     * @param pre The pre-hunt IntakeMessage
+     * @param post The post-hunt IntakeMessage
+     * @returns An array of keys (of IntakeMessage) for which properties has been exempted or null. 
+     */
+    getExemptions(pre: IntakeMessage, post: IntakeMessage): (keyof IntakeMessage)[] | null;
+}
+
+/**
+ * Base class for all rules to gather class name for logging
```

Comment:
Why does it need to be a base class rather than an interface? I'd prefer avoiding an inheritance hierarchy for something as trivial as "log the class name"

Reply:
It was to cut down on the amount of lines needed to write a rule/exemption. Without a base class to set the class name, each rule needed to implement the field.

```ts
export class UserNumActiveTurnsIncrementedByOne extends RuleBase<User> {
    isValid(pre: User, post: User): boolean {
        return post.num_active_turns - pre.num_active_turns === 1;
    }
}
```
or
```ts
export class UserNumActiveTurnsIncrementedByOne implements IRule<User> {
    name: string = "UserNumActiveTurnsIncrementedByOne"
    isValid(pre: User, post: User): boolean {
        return post.num_active_turns - pre.num_active_turns === 1;
    }
}
```
or as you mentioned in another comment override `toString` 
Downside to this: even if you add `toString` to the `IRule` interface, the compiler doesn't force any classes to implement it since it's technically already implemented by Object.
```ts
export class UserNumActiveTurnsIncrementedByOne implements IRule<User> {
    isValid(pre: User, post: User): boolean {
        return post.num_active_turns - pre.num_active_turns === 1;
    }
    
    toString(): string {
        return "UserNumActiveTurnsIncrementedByOne";
    }
}
```

I'm good with swapping to option 2 (and naming it like `description`). I wanted debug logging to be able to log why rules failed and each rule provide a short description of what it does. The class name did that but it's a little ugly.
But I don't think `toString` is enforceable if a new user were to create an exemption. I want the compiler to complain, not reviewers.

Reply:
Does anyone have a good suggestion for what the field should be named for debug logging?

Reply:
what are the field values?

Reply:
I've removed the inheritance and went with option 2. Renamed the `name` field to `description`. It's not too verbose.

@DevJackSmith The field value previously was the rule name, it was being set by the base class.
Now it's called 'description' and gives a very brief statement on what the rule/exemption does.